File: src/workspaceforge/commands/handoff.ts
import path from "node:path";
import { detectPartialState } from "../state/partial-state.js";
import { writeHandoff } from "../scaffold/handoff-generator.js";
import { EXIT_CODES } from "../util/exit-codes.js";

export interface HandoffNewCommandOptions {
  path?: string;
  json?: boolean;
}

export interface HandoffNewCommandOutcome {
  exitCode: number;
  json: Record<string, unknown>;
  humanLines: string[];
}

export async function runHandoffNewCommand(
  message: string,
  options: HandoffNewCommandOptions,
): Promise<HandoffNewCommandOutcome> {
  const repoPath = path.resolve(options.path ?? process.cwd());
  const workspaceDir = path.join(repoPath, ".workspace");

  if (!message || message.trim().length === 0) {
    return {
      exitCode: EXIT_CODES.GENERAL_ERROR,
      json: { ok: false, error: "empty_message", message: "handoff new requires a non-empty message." },
      humanLines: ["Error: handoff new requires a non-empty message."],
    };
  }

  const partialState = await detectPartialState(workspaceDir);
  if (partialState.type === "none") {
    return {
      exitCode: EXIT_CODES.NO_WORKSPACE_FOUND,
      json: {
        ok: false,
        error: "no_workspace_found",
        message: "No .workspace/ directory found. Run 'workspaceforge init' first.",
      },
      humanLines: ["No .workspace/ directory found. Run 'workspaceforge init' first."],
    };
  }

  const written = await writeHandoff(workspaceDir, repoPath, message.trim());

  return {
    exitCode: EXIT_CODES.OK,
    json: {
      ok: true,
      file: written.fileName,
      path: written.filePath,
      message: written.message,
      metadata: written.metadata,
    },
    humanLines: [`Handoff written: .workspace/handoff/${written.fileName}`],
  };
}

File: python/src/agenticworkspace/commands/handoff.py
"""
`agenticworkspace handoff new <message>` -- write a new timestamped session
handoff file. Ported from src/agenticworkspace/commands/handoff.ts.
"""
from __future__ import annotations

import os
from typing import Optional

from ..scaffold.handoff_generator import write_handoff
from ..state.partial_state import detect_partial_state
from ..util.exit_codes import EXIT_CODES
from .init import CommandOutcome


class HandoffNewCommandOptions:
    def __init__(self, path: Optional[str] = None, json: bool = False) -> None:
        self.path = path
        self.json = json


def run_handoff_new_command(message: str, options: HandoffNewCommandOptions) -> CommandOutcome:
    repo_path = os.path.abspath(options.path or os.getcwd())
    workspace_dir = os.path.join(repo_path, ".workspace")

    if not message or len(message.strip()) == 0:
        return CommandOutcome(
            exit_code=EXIT_CODES["GENERAL_ERROR"],
            json={"ok": False, "error": "empty_message", "message": "handoff new requires a non-empty message."},
            human_lines=["Error: handoff new requires a non-empty message."],
        )

    partial_state = detect_partial_state(workspace_dir)
    if partial_state.type == "none":
        return CommandOutcome(
            exit_code=EXIT_CODES["NO_WORKSPACE_FOUND"],
            json={
                "ok": False,
                "error": "no_workspace_found",
                "message": "No .workspace/ directory found. Run 'agenticworkspace init' first.",
            },
            human_lines=["No .workspace/ directory found. Run 'agenticworkspace init' first."],
        )

    written = write_handoff(workspace_dir, repo_path, message.strip())

    return CommandOutcome(
        exit_code=EXIT_CODES["OK"],
        json={
            "ok": True,
            "file": written.file_name,
            "path": written.file_path,
            "message": written.message,
            "metadata": {
                "timestamp": written.metadata.timestamp,
                "branch": written.metadata.branch,
                "commit": written.metadata.commit,
            },
        },
        human_lines=[f"Handoff written: .workspace/handoff/{written.file_name}"],
    )
